extension/httpfs/crypto.cpp

```diff
@@ -24,4 +30,118 @@ void hex256(hash_bytes &in, hash_str &out) {
 		pout[1] = hex[*pin & 0xF];
 	}
 }
+
+const EVP_CIPHER *GetCipher(const string &key) {
+	// For now, only ciphers for GCM Mode
+	switch (key.size()) {
+	case 16:
+		return EVP_aes_128_gcm();
+	case 24:
+		return EVP_aes_192_gcm();
+	case 32:
+		return EVP_aes_256_gcm();
+	default:
+		throw InternalException("Invalid AES key length");
+	}
+}
+
+AESGCMStateSSL::AESGCMStateSSL() : gcm_context(EVP_CIPHER_CTX_new()) {
+	auto context = reinterpret_cast<evp_cipher_ctx_st *>(gcm_context);
+	if (!(context)) {
+		throw InternalException("AES GCM failed with initializing context");
+	}
+}
+
+AESGCMStateSSL::~AESGCMStateSSL() {
+	// Clean up
+	EVP_CIPHER_CTX_free(reinterpret_cast<evp_cipher_ctx_st *>(gcm_context));
+}
+
+bool AESGCMStateSSL::IsOpenSSL() {
+	return ssl;
+}
+
+void AESGCMStateSSL::GenerateRandomData(data_ptr_t data, idx_t len) {
+	// generate random bytes for nonce
+	RAND_bytes(data, len);
+}
+
+void AESGCMStateSSL::InitializeEncryption(const_data_ptr_t iv, idx_t iv_len, const string *key) {
+	auto context = reinterpret_cast<evp_cipher_ctx_st *>(gcm_context);
+	mode = ENCRYPT;
+
+	if (1 != EVP_EncryptInit_ex(context, GetCipher(*key), NULL, const_data_ptr_cast(key->data()), iv)) {
+		throw InternalException("AES CTR failed with EncryptInit");
+	}
+}
+
+void AESGCMStateSSL::InitializeDecryption(const_data_ptr_t iv, idx_t iv_len, const string *key) {
+	auto context = reinterpret_cast<evp_cipher_ctx_st *>(gcm_context);
+	mode = DECRYPT;
+
+	if (1 != EVP_DecryptInit_ex(context, GetCipher(*key), NULL, const_data_ptr_cast(key->data()), iv)) {
+		throw std::runtime_error("EVP_EncryptInit_ex failed");
```

Comment:
Is there a reason this is a std::runtime_error instead of a DuckDB exception?

Reply:
No, thanks for pointing out!